### ISSUE TITLE
EXC_BAD_ACCESS when logging exceptions in http callback

### ISSUE THREAD

**Issue by steaks** [bugsnag/bugsnag-cocoa#115]
KSCrash hits EXC_BAD_ACCESS errors when  when logging exceptions inside an Optional callback invoked inside of `NSURLSession.sharedSession().dataTaskWithRequest`.  [DataTaskWithRequest](https://developer.apple.com/library/mac/documentation/Foundation/Reference/NSURLSession_class/#//apple_ref/occ/instm/NSURLSession/dataTaskWithRequest:completionHandler:) is a common component used for http requests, so this issue may be seen in a lot of apps.  I stumbled across this issue my first time writing a wrapper function for GET requests.

Furthermore, after hitting this error one time, the app will crash every successive time Bugsnag is setup.  This means the **app will never work again** after a user hits this issue once.

How to reproduce:

```
class MyClass {
    class func httpGetBad(url: NSURL, completionHandler: ((response: NSURLResponse?, data: NSData?, error: NSError?) -> Void)?) {
        let request = NSMutableURLRequest(URL: url)
        request.HTTPMethod = "GET"
        let task = NSURLSession.sharedSession().dataTaskWithRequest(request) { data, response, error in
            if let callback = completionHandler {
                callback(response: response, data: data, error: error) //Any error logged in this callback will make the app unusable forever
            }
        }
        task.resume()
    }
    
    class func httpGetGood(url: NSURL, completionHandler: (response: NSURLResponse?, data: NSData?, error: NSError?) -> Void) {
        let request = NSMutableURLRequest(URL: url)
        request.HTTPMethod = "GET"
        let task = NSURLSession.sharedSession().dataTaskWithRequest(request) { data, response, error in
            let callback = completionHandler
            callback(response: response, data: data, error: error)
        }
        task.resume()
    }
}
```

```
let url = NSURL(string: "https://www.google.com")!
httpGetBad(url, completionHandler: { (response, data, error) in
  let exception = NSException(name: "Test", reason: "Test reason", userInfo: nil)
  Bugsnag.notify(exception) //Now the app can never be used again!
})
```

```
Bugsnag.startBugsnagWithConfiguration(config) //Will hit an EXC_BAD_ACCESS every time now!
```

<br/>
_Reproduced consistently on iPhone 6, iOS 9.2.1, Bugsnag version 5.4.2_

**Comment by kattrali** [comment]
Thank you for the report, @steaks. Can you confirm that this is on the latest version, v5.4.2? I’ll investigate this further.

**Comment by steaks** [comment]
@kattrali updated the issue.  This was reproduced on 5.4.2.

**Comment by kattrali** [comment]
Thanks for the update, @steaks. I added the sample code to the example swift project in https://github.com/bugsnag/bugsnag-cocoa/commit/d67e10a6b012db5a393ea575bcaceece1f04d5b0 on branch [`kattrali/test-crash-115`](https://github.com/bugsnag/bugsnag-cocoa/tree/kattrali/test-crash-115). Can you confirm by running `examples/swift-ios/bugsnag-example.xcworkspace` that the sample code reproduces the problem?

I have not successfully reproduced with:

* Bugsnag 5.4.2
* KSCrash 1.6.4
* Xcode 7.3.1
* iOS 9.3.4 / iPhone SE

though I can try with an iPhone 6 later today.

**Comment by steaks** [comment]
Hey @kattrali I had some trouble reproducing the issue also.  I think it's because I upgraded my xcode from 7.2.1 to 7.3.1 this morning.  We're honing in on what is required to reproduce.  I think it's:

- Bugsnag 5.4.2
- KSCrash 1.6.4
- XCode 7.2.1
- iOS 9.2.1 (not simulator)


One of our team memebers still has XCode 7.2.1, so he's going to test today or tomorrow.

P.S. I had trouble building the projects in the Bugsnag repo.  So I'm using this repository to test: https://github.com/steaks/bugsnag-test-two

**Comment by kattrali** [comment]
Thanks @steaks. You probably needed to `git pull --recursive` to build, as the main repo uses submodules to handle the dependencies. I’ll take a look at the fork.

**Comment by kattrali** [comment]
@steaks Have you had any success reproducing the issue? I haven’t seen it occur yet using Xcode 7.2.1 and iOS 9.2.1, though I haven’t been able to test more than the iPhone SE and iPhone 6.

**Comment by kattrali** [comment]
After some spelunking, this was an issue in the event storage logic in [KSCrash 1.6.4 KSZombie.m](https://github.com/kstenerud/KSCrash/blob/1.6.4/Source/KSCrash/Recording/Tools/KSZombie.m#L126), which has been updated and corrected in [KSCrash 1.8.5](https://github.com/kstenerud/KSCrash/commit/dded5fd025eb84a47f972795c252a5dc199feb4f).  I've released a new version of Bugsnag requiring at least 1.8.5 as the minimum version.
